src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Performance/MemoryCacheBenchmark.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using BenchmarkDotNet.Attributes;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.Serialization;
+using Newtonsoft.Json;
+
+namespace Microsoft.AspNetCore.Razor.Performance
+{
+    public class MemoryCacheBenchmark
+    {
+        public MemoryCacheBenchmark()
+        {
+            var current = new DirectoryInfo(AppContext.BaseDirectory);
+            while (current != null && !File.Exists(Path.Combine(current.FullName, "taghelpers.json")))
+            {
+                current = current.Parent;
+            }
+
+            var tagHelperFilePath = Path.Combine(current.FullName, "taghelpers.json");
+            var tagHelperBuffer = File.ReadAllBytes(tagHelperFilePath);
+
+            // Deserialize from json file.
+            var serializer = new JsonSerializer();
+            serializer.Converters.Add(new TagHelperDescriptorJsonConverter());
+            using (var stream = new MemoryStream(tagHelperBuffer))
+            using (var reader = new JsonTextReader(new StreamReader(stream)))
+            {
+                TagHelpers = serializer.Deserialize<IReadOnlyList<TagHelperDescriptor>>(reader);
+                TagHelperHashes = TagHelpers.Select(th => th.GetHashCode()).ToList();
+            }
+
+            // Set cache size to 400 so anything more then that will force compacts
+            Cache = new MemoryCache<int, TagHelperDescriptor>(400);
+        }
+
+        private IReadOnlyList<int> TagHelperHashes { get; }
+
+        private IReadOnlyList<TagHelperDescriptor> TagHelpers { get; }
+
+        private MemoryCache<int, TagHelperDescriptor> Cache { get; }
+
+        [Benchmark(Description = "MemoryCache Set performance with limited size")]
+        public void Set_Performance()
+        {
+            for (var i = 0; i < TagHelpers.Count; i++)
+            {
+                Cache.Set(TagHelperHashes[i], TagHelpers[i]);
```

Comment:
>Cache.Set(TagHelperHashes[i], TagHelpers[i]); [](start = 16, length = 45)

Out of curiosity, couldn't this have just been benchmarked without the taghelper infrastructure, ie just 

private MemoryCache<int, int> Cache { get; }

public MemoryCacheBenchmark()
{
    Cache = new MemoryCache<int, int>(400);
}

[Benchmark(Description = "MemoryCache Set performance with limited size")]
public void Set_Performance()
{
    for (var i = 0; i < 1000; i++)
    {
        Cache.Set(i, i);
    }
}

Reply:
TagHelperDescriptor's are IEqutable so I believe the Remove call in `Compact` utilizes that. That was my intent anyhow.

Reply:
I just would have thought that the additional overhead/complexity of using taghelpers to measure the memorycache improvements wouldn't be necessary.

Reply:
> I just would have thought that the additional overhead/complexity of using taghelpers to measure the memorycache improvements wouldn't be necessary.

Agree but the end-user scenario that was breaking was one that was using TagHelpers so wanted to be extra sure. I also didn't want to rule out that the complexity of the [TagHelperDescriptor GetHashCode/Equals](https://github.com/dotnet/aspnetcore/blob/277cc564ce5a76c6613f29dd92c0168e06b41b1c/src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorComparer.cs#L22-L167) comparisons was at fault either (which I'm sure it definitely contributed).

If I were to change it now to utilize something different other then TagHelperDescriptors I'd be testing the performance of the cache in a way that wasn't actually used in production.